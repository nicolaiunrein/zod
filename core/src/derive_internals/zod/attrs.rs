use darling::{FromAttributes, FromDeriveInput};
use syn::{Expr, Lit, Meta};

use super::{fields::FieldValue, Derive};

#[derive(FromDeriveInput)]
#[darling(
    attributes(zod),
    forward_attrs(allow, doc, cfg),
    supports(struct_named, struct_newtype, struct_tuple, enum_any)
)]

pub(super) struct ZodAttrs {
    pub namespace: syn::Path,
    pub custom_suffix: Option<String>,
}

#[derive(FromAttributes)]
#[darling(attributes(zod))]
pub(super) struct ZodFieldAttrs {
    override_input_with: Option<syn::Path>,
    override_output_with: Option<syn::Path>,
    override_with: Option<syn::Path>,
}

impl ZodFieldAttrs {
    pub(super) fn as_field_value(&self, derive: Derive) -> Option<FieldValue> {
        let fail = || match derive {
            Derive::Input => {
                r#"only one of #[zod(override_with = "...")] or #[zod(override_input_with = "...")] may be specified"#
            }
            Derive::Output => {
                r#"only one of #[zod(override_with = "...")] or #[zod(override_output_with = "...")] may be specified"#
            }
        };
        match derive {
            Derive::Input => match (&self.override_with, &self.override_input_with) {
                (None, None) => None,
                (None, Some(p)) | (Some(p), None) => Some(FieldValue::OverrideGetter(p.clone())),
                _ => panic!("{}", fail()),
            },
            Derive::Output => match (&self.override_with, &self.override_output_with) {
                (None, None) => None,
                (None, Some(p)) | (Some(p), None) => Some(FieldValue::OverrideGetter(p.clone())),
                _ => panic!("{}", fail()),
            },
        }
    }
}

pub trait FieldAttrsExt {
    fn skip(&self, derive: Derive) -> bool;
    fn is_optional(&self, derive: Derive) -> bool;
}

impl FieldAttrsExt for serde_derive_internals::attr::Field {
    fn skip(&self, derive: Derive) -> bool {
        match derive {
            Derive::Input => self.skip_deserializing(),
            Derive::Output => self.skip_serializing(),
        }
    }

    fn is_optional(&self, derive: Derive) -> bool {
        match derive {
            Derive::Input => !self.default().is_none(),
            Derive::Output => !self.default().is_none() || self.skip_serializing_if().is_some(),
        }
    }
}

pub(crate) trait NameExt {
    fn as_str(&self, derive: Derive) -> String;
}

impl NameExt for serde_derive_internals::attr::Name {
    fn as_str(&self, derive: Derive) -> String {
        match derive {
            Derive::Input => self.deserialize_name(),
            Derive::Output => self.serialize_name(),
        }
    }
}

pub fn get_rustdoc(attrs: &[syn::Attribute]) -> Option<String> {
    let mut full_docs = String::new();
    for attr in attrs {
        match attr.meta {
            Meta::NameValue(ref nv) => {
                if nv.path.is_ident("doc") {
                    match nv.value {
                        Expr::Lit(ref lit_expr) => match lit_expr.lit {
                            Lit::Str(ref s) => {
                                full_docs.push_str(s.value().trim());
                                full_docs.push('\n');
                            }
                            _ => {}
                        },
                        _ => {}
                    }
                }
            }
            _ => {}
        }
    }
    if full_docs.is_empty() {
        None
    } else {
        Some(full_docs)
    }
}
