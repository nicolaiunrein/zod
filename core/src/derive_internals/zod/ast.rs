use super::attrs::{FieldAttrsExt, NameExt};
use super::custom_suffix::CustomSuffix;
use super::fields::{FieldValue, ZodNamedFieldImpl, ZodUnnamedFieldImpl};
use super::generics::{needs_inline, replace_generics, GenericsExt};
use super::r#enum::{EnumImpl, TagType, VariantImpl};
use super::r#struct::{ZodObjectImpl, ZodTupleImpl};
use super::Derive;
use super::{attrs::ZodAttrs, custom_suffix};
use crate::utils::zod_core;
use crate::Kind;
use darling::FromDeriveInput;
use proc_macro2::TokenStream as TokenStream2;
use quote::{quote, quote_spanned, ToTokens};
use syn::DeriveInput;

pub(super) struct Ast {
    derive: Derive,
    pub ident: syn::Ident,
    pub data: Data,
    pub generics: syn::Generics,
    pub namespace: syn::Path,
    pub custom_suffix: custom_suffix::CustomSuffix,
    pub name: String,
    pub optional: bool,
    // pub transparent: bool,
    // pub type_from: Option<syn::Type>,
    // pub type_try_from: Option<syn::Type>,
    // pub type_into: Option<syn::Type>,
}

impl Ast {
    pub fn new(derive: Derive, mut derive_input: DeriveInput) -> Result<Self, TokenStream2> {
        let cx = serde_derive_internals::Ctxt::new();
        let input_clone = derive_input.clone();
        let serde_ast =
            serde_derive_internals::ast::Container::from_ast(&cx, &input_clone, derive.into())
                .unwrap();

        cx.check().unwrap();

        let zod_attrs: ZodAttrs = match ZodAttrs::from_derive_input(&derive_input) {
            Ok(attrs) => attrs,
            Err(err) => return Err(err.write_errors()),
        };

        derive_input.generics.update_where_clause(derive);

        let name = match derive {
            Derive::Input => serde_ast.attrs.name().deserialize_name(),
            Derive::Output => serde_ast.attrs.name().serialize_name(),
        };

        let mut custom_suffix = CustomSuffix {
            inner: zod_attrs.custom_suffix,
        };

        if serde_ast.attrs.deny_unknown_fields() {
            custom_suffix.add(".strict()");
        }

        Ok(Self {
            derive,
            ident: derive_input.ident.clone(),
            optional: !serde_ast.attrs.default().is_none(),
            data: Data::new(derive, serde_ast),
            generics: derive_input.generics,
            namespace: zod_attrs.namespace,
            custom_suffix,
            name,
        })
    }

    fn generic_arguments(&self) -> Vec<TokenStream2> {
        self.generics
            .idents()
            .iter()
            .map(|ident| {
                let name = ident.to_string();

                quote_spanned! {
                    ident.span() =>
                    #zod_core::GenericArgument::new::<#ident>(#name)
                }
            })
            .collect()
    }

    fn unique_ident(&self) -> syn::Ident {
        let name = &self.name;
        match self.derive {
            Derive::Input => {
                crate::utils::make_unique_name::<Kind::Input>(&quote::format_ident!("{name}"))
            }
            Derive::Output => {
                crate::utils::make_unique_name::<Kind::Output>(&quote::format_ident!("{name}"))
            }
        }
    }
}

impl ToTokens for Ast {
    fn to_tokens(&self, tokens: &mut TokenStream2) {
        let ident = &self.ident;
        let ns = &self.namespace;
        let name = &self.name;
        let custom_suffix = &self.custom_suffix;
        let inline = self.data.inline();
        let inner = &self.data;
        let generic_arguments = self.generic_arguments();
        let unique_ident = self.unique_ident();
        let derive = self.derive;
        let optional = self.optional;

        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();

        tokens.extend(quote! {
            impl #impl_generics #zod_core::Type<#derive> for #ident #ty_generics #where_clause {
                type Ns = #ns;
                const NAME: &'static str = #name;
                const INLINE: bool = #inline;

                fn value() -> #zod_core::z::ZodType<#derive> {
                    #zod_core::z::ZodType {
                        optional: #optional,
                        custom_suffix: #custom_suffix,
                        inner: #inner.into()
                    }
                }

                fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                    let mut v = ::std::vec::Vec::new();
                    #(v.push(#generic_arguments);)*
                    v
                }

                fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#derive>) {
                    // TODO
                }
            }

            impl #ns {
                #[allow(dead_code)]
                #[allow(non_upper_case_globals)]
                const #unique_ident: () = {};
            }
        })
    }
}

#[derive(Debug, PartialEq, Clone)]
pub(super) enum Data {
    Struct(bool, ZodObjectImpl),
    Tuple(bool, ZodTupleImpl),
    Enum(bool, EnumImpl),
}

impl ToTokens for Data {
    fn to_tokens(&self, tokens: &mut TokenStream2) {
        match self {
            Data::Struct(_, inner) => inner.to_tokens(tokens),
            Data::Tuple(_, inner) => inner.to_tokens(tokens),
            Data::Enum(_, inner) => inner.to_tokens(tokens),
        }
    }
}

impl Data {
    fn new(derive: Derive, input: serde_derive_internals::ast::Container) -> Self {
        match input.data {
            serde_derive_internals::ast::Data::Struct(style, fields) => {
                let inline = fields
                    .iter()
                    .filter(|field| !field.attrs.skip(derive))
                    .any(|f| needs_inline(&f.ty, &input.generics));

                match style {
                    serde_derive_internals::ast::Style::Struct => Self::Struct(
                        inline,
                        ZodObjectImpl {
                            fields: fields
                                .into_iter()
                                .filter(|field| !field.attrs.skip(derive))
                                .map(|field| {
                                    let mut ty = field.ty.clone();
                                    if !inline {
                                        replace_generics(&mut ty, &input.generics);
                                    }
                                    ZodNamedFieldImpl {
                                        name: field.attrs.name().as_str(derive),
                                        optional: field.attrs.is_optional(derive),
                                        derive,
                                        value: FieldValue::Type(ty),
                                    }
                                })
                                .collect(),
                        },
                    ),
                    serde_derive_internals::ast::Style::Newtype
                    | serde_derive_internals::ast::Style::Tuple => Self::Tuple(
                        inline,
                        ZodTupleImpl {
                            fields: fields
                                .into_iter()
                                .filter(|field| !field.attrs.skip(derive))
                                .map(|field| {
                                    let mut ty = field.ty.clone();
                                    if !inline {
                                        replace_generics(&mut ty, &input.generics);
                                    }
                                    ZodUnnamedFieldImpl {
                                        derive,
                                        optional: field.attrs.is_optional(derive),
                                        ty,
                                    }
                                })
                                .collect(),
                        },
                    ),
                    serde_derive_internals::ast::Style::Unit => todo!(),
                }
            }
            serde_derive_internals::ast::Data::Enum(variants) => {
                let inline = variants.iter().any(|v| {
                    v.fields
                        .iter()
                        .filter(|field| !field.attrs.skip(derive))
                        .any(|f| needs_inline(&f.ty, &input.generics))
                });

                let variants: Vec<_> = variants
                    .into_iter()
                    .map(|variant| match input.attrs.tag().into() {
                        TagType::Externally => match variant.style {
                            serde_derive_internals::ast::Style::Unit => {
                                VariantImpl::Literal(variant.attrs.name().as_str(derive))
                            }
                            serde_derive_internals::ast::Style::Struct => {
                                VariantImpl::Object(ZodObjectImpl {
                                    fields: vec![ZodNamedFieldImpl {
                                        name: variant.attrs.name().as_str(derive),
                                        optional: false,
                                        derive,
                                        value: FieldValue::Object(ZodObjectImpl {
                                            fields: variant
                                                .fields
                                                .into_iter()
                                                .filter(|field| !field.attrs.skip(derive))
                                                .map(|f| {
                                                    let mut ty = f.ty.clone();
                                                    if !inline {
                                                        replace_generics(&mut ty, &input.generics);
                                                    }

                                                    ZodNamedFieldImpl {
                                                        name: f.attrs.name().as_str(derive),
                                                        optional: f.attrs.is_optional(derive),
                                                        derive,
                                                        value: FieldValue::Type(ty),
                                                    }
                                                })
                                                .collect(),
                                        }),
                                    }],
                                })
                            }
                            serde_derive_internals::ast::Style::Newtype
                            | serde_derive_internals::ast::Style::Tuple => {
                                VariantImpl::Object(ZodObjectImpl {
                                    fields: vec![ZodNamedFieldImpl {
                                        name: variant.attrs.name().as_str(derive),
                                        optional: false,
                                        derive,
                                        value: FieldValue::Tuple(ZodTupleImpl {
                                            fields: variant
                                                .fields
                                                .into_iter()
                                                .filter(|field| !field.attrs.skip(derive))
                                                .map(|f| {
                                                    let mut ty = f.ty.clone();
                                                    if !inline {
                                                        replace_generics(&mut ty, &input.generics);
                                                    }

                                                    ZodUnnamedFieldImpl {
                                                        optional: f.attrs.is_optional(derive),
                                                        derive,
                                                        ty,
                                                    }
                                                })
                                                .collect(),
                                        }),
                                    }],
                                })
                            }
                        },
                        TagType::Internally { tag } => match variant.style {
                            serde_derive_internals::ast::Style::Unit => {
                                VariantImpl::Object(ZodObjectImpl {
                                    fields: vec![ZodNamedFieldImpl {
                                        name: tag,
                                        optional: false,
                                        derive,
                                        value: FieldValue::Literal(
                                            variant.attrs.name().as_str(derive),
                                            variant.ident.span(),
                                        ),
                                    }],
                                })
                            }
                            serde_derive_internals::ast::Style::Struct => {
                                let first = ZodNamedFieldImpl {
                                    name: tag.clone(),
                                    optional: false,
                                    derive,
                                    value: FieldValue::Literal(
                                        variant.attrs.name().as_str(derive),
                                        variant.ident.span(),
                                    ),
                                };
                                let fields = variant
                                    .fields
                                    .iter()
                                    .filter(|field| !field.attrs.skip(derive))
                                    .map(|f| {
                                        let mut ty = f.ty.clone();

                                        if !inline {
                                            replace_generics(&mut ty, &input.generics)
                                        }

                                        ZodNamedFieldImpl {
                                            name: f.attrs.name().as_str(derive),
                                            optional: f.attrs.is_optional(derive),
                                            derive,
                                            value: FieldValue::Type(ty),
                                        }
                                    });

                                VariantImpl::Object(ZodObjectImpl {
                                    fields: std::iter::once(first).chain(fields).collect(),
                                })
                            }
                            serde_derive_internals::ast::Style::Tuple => {
                                unreachable!("prevented by serde")
                            }
                            serde_derive_internals::ast::Style::Newtype => {
                                todo!("Serde supports object merging")
                            }
                        },
                        TagType::Adjacently { tag, content } => match variant.style {
                            serde_derive_internals::ast::Style::Unit => {
                                VariantImpl::Object(ZodObjectImpl {
                                    fields: vec![ZodNamedFieldImpl {
                                        name: tag,
                                        optional: false,
                                        derive,
                                        value: FieldValue::Literal(
                                            variant.attrs.name().as_str(derive),
                                            variant.ident.span(),
                                        ),
                                    }],
                                })
                            }
                            serde_derive_internals::ast::Style::Struct => {
                                let value = FieldValue::Object(ZodObjectImpl {
                                    fields: variant
                                        .fields
                                        .iter()
                                        .filter(|field| !field.attrs.skip(derive))
                                        .map(|f| {
                                            let mut ty = f.ty.clone();
                                            if !inline {
                                                replace_generics(&mut ty, &input.generics)
                                            }

                                            ZodNamedFieldImpl {
                                                name: f.attrs.name().as_str(derive),
                                                optional: f.attrs.is_optional(derive),
                                                derive,
                                                value: FieldValue::Type(f.ty.clone()),
                                            }
                                        })
                                        .collect(),
                                });

                                VariantImpl::Object(ZodObjectImpl {
                                    fields: vec![
                                        ZodNamedFieldImpl {
                                            name: tag,
                                            optional: false,
                                            derive,
                                            value: FieldValue::Literal(
                                                variant.attrs.name().as_str(derive),
                                                variant.ident.span(),
                                            ),
                                        },
                                        ZodNamedFieldImpl {
                                            name: content,
                                            optional: false,
                                            derive,
                                            value,
                                        },
                                    ],
                                })
                            }
                            serde_derive_internals::ast::Style::Newtype
                            | serde_derive_internals::ast::Style::Tuple => {
                                let value = FieldValue::Tuple(ZodTupleImpl {
                                    fields: variant
                                        .fields
                                        .into_iter()
                                        .filter(|field| !field.attrs.skip(derive))
                                        .map(|f| {
                                            let mut ty = f.ty.clone();
                                            if !inline {
                                                replace_generics(&mut ty, &input.generics)
                                            }

                                            ZodUnnamedFieldImpl {
                                                derive,
                                                optional: f.attrs.is_optional(derive),
                                                ty,
                                            }
                                        })
                                        .collect(),
                                });

                                VariantImpl::Object(ZodObjectImpl {
                                    fields: vec![
                                        ZodNamedFieldImpl {
                                            name: tag,
                                            optional: false,
                                            derive,
                                            value: FieldValue::Literal(
                                                variant.attrs.name().as_str(derive),
                                                variant.ident.span(),
                                            ),
                                        },
                                        ZodNamedFieldImpl {
                                            name: content,
                                            optional: false,
                                            derive,
                                            value,
                                        },
                                    ],
                                })
                            }
                        },
                        TagType::Untagged => match variant.style {
                            serde_derive_internals::ast::Style::Struct => {
                                VariantImpl::Object(ZodObjectImpl {
                                    fields: variant
                                        .fields
                                        .into_iter()
                                        .filter(|field| !field.attrs.skip(derive))
                                        .map(|f| {
                                            let mut ty = f.ty.clone();
                                            if !inline {
                                                replace_generics(&mut ty, input.generics)
                                            }
                                            ZodNamedFieldImpl {
                                                name: f.attrs.name().as_str(derive),
                                                optional: f.attrs.is_optional(derive),
                                                derive,
                                                value: FieldValue::Type(ty),
                                            }
                                        })
                                        .collect(),
                                })
                            }
                            serde_derive_internals::ast::Style::Newtype
                            | serde_derive_internals::ast::Style::Tuple => {
                                VariantImpl::Tuple(ZodTupleImpl {
                                    fields: variant
                                        .fields
                                        .into_iter()
                                        .filter(|field| !field.attrs.skip(derive))
                                        .map(|f| {
                                            let mut ty = f.ty.clone();
                                            if !inline {
                                                replace_generics(&mut ty, input.generics)
                                            }
                                            ZodUnnamedFieldImpl {
                                                optional: f.attrs.is_optional(derive),
                                                derive,
                                                ty,
                                            }
                                        })
                                        .collect(),
                                })
                            }
                            serde_derive_internals::ast::Style::Unit => {
                                VariantImpl::Literal(variant.attrs.name().as_str(derive))
                            }
                        },
                    })
                    .collect();

                Self::Enum(
                    inline,
                    EnumImpl {
                        tag: input.attrs.tag().into(),
                        variants,
                        derive,
                    },
                )
            }
        }
    }

    fn inline(&self) -> bool {
        match self {
            Data::Struct(inline, _) => *inline,
            Data::Tuple(inline, _) => *inline,
            Data::Enum(inline, _) => *inline,
        }
    }
}

#[cfg(test)]
mod test {
    use crate::derive_internals::zod::fields::{FieldValue, ZodNamedFieldImpl};

    use super::*;
    use pretty_assertions::assert_eq;
    use syn::parse_quote;

    #[test]
    fn non_inline_generics_ok() {
        let input = parse_quote! {
            struct X<T> {
                inner: Other<T>
            }
        };

        let cx = serde_derive_internals::Ctxt::new();
        let input = serde_derive_internals::ast::Container::from_ast(
            &cx,
            &input,
            serde_derive_internals::Derive::Serialize,
        )
        .unwrap();
        cx.check().unwrap();

        let data = Data::new(Derive::Input, input);

        assert_eq!(
            data,
            Data::Struct(
                false,
                ZodObjectImpl {
                    fields: vec![ZodNamedFieldImpl {
                        name: String::from("inner"),
                        optional: false,
                        derive: Derive::Input,
                        value: FieldValue::Type(
                            parse_quote!(Other<#zod_core::typed_str::TypedStr<'T', #zod_core::typed_str::End>>),
                        )
                    }]
                }
            )
        )
    }

    #[test]
    fn inline_generics_ok() {
        let input = parse_quote! {
            struct X<T: SomeTrait> {
                inner: Other<T>
            }
        };

        let cx = serde_derive_internals::Ctxt::new();
        let input = serde_derive_internals::ast::Container::from_ast(
            &cx,
            &input,
            serde_derive_internals::Derive::Serialize,
        )
        .unwrap();
        cx.check().unwrap();

        let data = Data::new(Derive::Input, input);

        assert_eq!(
            data,
            Data::Struct(
                true,
                ZodObjectImpl {
                    fields: vec![ZodNamedFieldImpl {
                        name: String::from("inner"),
                        optional: false,
                        derive: Derive::Input,
                        value: FieldValue::Type(parse_quote!(Other<T>))
                    }]
                }
            )
        )
    }
}
