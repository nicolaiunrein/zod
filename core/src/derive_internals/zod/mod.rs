mod ast;
mod attrs;
mod custom_suffix;
mod data;
mod r#enum;
mod fields;
mod generics;
mod r#struct;
mod variant;

use self::ast::Ast;
use crate::utils::zod_core;
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use serde_derive_internals::Derive as SerdeDerive;
use syn::DeriveInput;

#[derive(Clone, PartialEq, Eq, Debug, Copy)]
pub enum Derive {
    Input,
    Output,
}

impl From<Derive> for SerdeDerive {
    fn from(value: Derive) -> Self {
        match value {
            Derive::Input => SerdeDerive::Deserialize,
            Derive::Output => SerdeDerive::Serialize,
        }
    }
}

impl ToTokens for Derive {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            Self::Input => tokens.extend(quote!(#zod_core::Kind::Input)),
            Self::Output => tokens.extend(quote!(#zod_core::Kind::Output)),
        }
    }
}

/// convert input into the generated code providing a `Derive`.
pub fn expand(derive: Derive, input: TokenStream) -> TokenStream {
    let derive_input: DeriveInput = match syn::parse2(input) {
        Ok(parsed) => parsed,
        Err(err) => {
            return err.into_compile_error();
        }
    };

    let ast = match Ast::new(derive, derive_input) {
        Ok(attrs) => attrs,
        Err(err) => return err,
    };

    ast.into_token_stream()
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::{
        derive_internals::zod::{
            custom_suffix::CustomSuffix,
            fields::{FieldValue, ZodNamedFieldImpl},
            r#struct::ZodObjectImpl,
        },
        test_utils::TokenStreamExt,
    };
    use pretty_assertions::assert_eq;
    use quote::quote;
    use syn::parse_quote;

    #[test]
    fn expand_ok() {
        let derive = Derive::Input;
        let input = quote! {
            #[zod(namespace = "Ns")]
            struct Test {
                inner_string: String,
                inner_u8: u8
            }
        };

        let inner = ZodObjectImpl {
            fields: vec![
                ZodNamedFieldImpl {
                    name: String::from("inner_string"),
                    optional: false,
                    derive,
                    value: FieldValue::Type(parse_quote!(String)),
                },
                ZodNamedFieldImpl {
                    name: String::from("inner_u8"),
                    optional: false,
                    derive,
                    value: FieldValue::Type(parse_quote!(u8)),
                },
            ],
        };

        let custom_suffix = CustomSuffix { inner: None };

        let expected = quote! {
            const _: () = {
                impl #zod_core::Type<#derive> for Test {
                    type Ns = Ns;
                    const NAME: &'static str = "Test";
                    const INLINE: bool = false;

                    fn value() -> #zod_core::z::ZodType<#derive> {
                        #zod_core::z::ZodType {
                            optional: false,
                            custom_suffix: #custom_suffix,
                            inner: #inner.into()
                        }
                    }

                    fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                        let mut v = ::std::vec::Vec::new();
                        v
                    }

                    fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#zod_core::Kind::Input>) {}
                }

                impl Ns {
                    #[allow(dead_code)]
                    #[allow(non_upper_case_globals)]
                    const __ZOD_PRIVATE_INPUT___Test: () = {};
                }
            };
        };

        assert_eq!(
            expand(Derive::Input, input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        )
    }
}
