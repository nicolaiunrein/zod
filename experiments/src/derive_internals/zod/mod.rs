mod ast;
mod attrs;
mod custom_suffix;
mod r#enum;
mod fields;
mod generics;
mod r#struct;

use self::ast::Ast;
use crate::utils::zod_core;
use proc_macro2::TokenStream as TokenStream2;
use quote::{quote, ToTokens};
use syn::DeriveInput;

#[derive(Clone, PartialEq, Eq, Debug, Copy)]
pub enum Derive {
    Input,
    Output,
}

impl ToTokens for Derive {
    fn to_tokens(&self, tokens: &mut TokenStream2) {
        match self {
            Self::Input => tokens.extend(quote!(#zod_core::Kind::Input)),
            Self::Output => tokens.extend(quote!(#zod_core::Kind::Output)),
        }
    }
}

/// convert input into the generated code providing a `Derive`.
pub fn expand(derive: Derive, input: TokenStream2) -> TokenStream2 {
    let derive_input: DeriveInput = match syn::parse2(input) {
        Ok(parsed) => parsed,
        Err(err) => {
            return err.into_compile_error();
        }
    };

    let ast = match Ast::new(derive, derive_input) {
        Ok(attrs) => attrs,
        Err(err) => return err,
    };

    ast.into_token_stream()
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::{
        derive_internals::zod::{
            custom_suffix::CustomSuffix,
            r#enum::EnumImpl,
            r#struct::{ZodObjectImpl, ZodTupleImpl},
        },
        test_utils::TokenStreamExt,
    };
    use pretty_assertions::assert_eq;
    use quote::quote;
    use syn::parse_quote;

    #[test]
    fn expand_zod_for_struct_with_named_fields_ok() {
        let derive = Derive::Input;
        let input = quote! {
            #[zod(namespace = "Ns")]
            struct Test {
                inner_string: String,
                inner_u8: u8
            }
        };

        let inner = ZodObjectImpl::new(
            derive,
            &parse_quote!({ inner_string: String, inner_u8: u8 }),
        );

        let custom_suffix = CustomSuffix { inner: None };

        let expected = quote! {
            impl #zod_core::Type<#derive> for Test {
                type Ns = Ns;
                const NAME: &'static str = "Test";
                const INLINE: bool = false;

                fn value() -> #zod_core::z::ZodType<#derive> {
                    #zod_core::z::ZodType {
                        optional: false,
                        custom_suffix: #custom_suffix,
                        inner: #inner.into()
                    }
                }

                fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                    let mut v = ::std::vec::Vec::new();
                    v
                }

                fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#zod_core::Kind::Input>) {}
            }

            impl Ns {
                #[allow(dead_code)]
                #[allow(non_upper_case_globals)]
                const __ZOD_PRIVATE_INPUT___Test: () = {};
            }

        };

        assert_eq!(
            expand(Derive::Input, input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        )
    }

    #[test]
    fn expand_zod_for_struct_with_tuple_fields_ok() {
        let derive = Derive::Input;
        let input = quote! {
            #[zod(namespace = "Ns")]
            struct Test(String, u8);
        };

        let custom_suffix = CustomSuffix { inner: None };

        let inner = ZodTupleImpl::new(derive, &parse_quote!((String, u8)));

        let expected = quote! {
            impl #zod_core::Type<#derive> for Test {
                type Ns = Ns;
                const NAME: &'static str = "Test";
                const INLINE: bool = false;

                fn value() -> #zod_core::z::ZodType<#derive> {
                    #zod_core::z::ZodType {
                        optional: false,
                        custom_suffix: #custom_suffix,
                        inner: #inner.into()
                    }
                }

                fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                    let mut v = ::std::vec::Vec::new();
                    v
                }

                fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#zod_core::Kind::Input>) {}
            }

            impl Ns {
                #[allow(dead_code)]
                #[allow(non_upper_case_globals)]
                const __ZOD_PRIVATE_INPUT___Test: () = {};
            }

        };

        assert_eq!(
            expand(Derive::Input, input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        )
    }

    #[test]
    fn impl_zod_for_enum() {
        let derive = Derive::Input;
        let input = quote! {
            #[zod(namespace = "Ns")]
            enum Test {
                Unit,
                Tuple1(String),
                Tuple2(String, u8),
                Struct0 {},
                Struct1 {
                    inner: String,
                },
                Struct2 {
                    inner_string: String,
                    inner_u8: u8,
                }
            }
        };

        let inner = EnumImpl::new(
            derive,
            Default::default(),
            vec![
                parse_quote!(Unit),
                parse_quote!(Tuple1(String)),
                parse_quote!(Tuple2(String, u8)),
                parse_quote!(Struct0 {}),
                parse_quote!(Struct1 { inner: String }),
                parse_quote!(Struct2 {
                    inner_string: String,
                    inner_u8: u8
                }),
            ],
        );

        let custom_suffix = CustomSuffix { inner: None };

        let expected = quote! {
            impl #zod_core::Type<#derive> for Test {
                type Ns = Ns;
                const NAME: &'static str = "Test";
                const INLINE: bool = false;

                fn value() -> #zod_core::z::ZodType<#derive> {
                    #zod_core::z::ZodType {
                        optional: false,
                        custom_suffix: #custom_suffix,
                        inner: #inner.into()
                    }
                }

                fn args() -> ::std::vec::Vec<#zod_core::GenericArgument<#derive>> {
                    let mut v = ::std::vec::Vec::new();
                    v
                }

                fn visit_dependencies(visitor: &mut #zod_core::DependencyVisitor<#zod_core::Kind::Input>) {}
            }

            impl Ns {
                #[allow(dead_code)]
                #[allow(non_upper_case_globals)]
                const __ZOD_PRIVATE_INPUT___Test: () = {};
            }

        };

        assert_eq!(
            expand(Derive::Input, input).to_formatted_string().unwrap(),
            expected.to_formatted_string().unwrap()
        );
    }
}
