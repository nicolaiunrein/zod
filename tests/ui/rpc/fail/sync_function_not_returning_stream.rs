use zod::{rpc, Namespace};

#[derive(Namespace)]
struct A;

#[rpc]
impl A {
    fn test(&mut self) -> String {
        String::new()
    }
}

fn main() {}
