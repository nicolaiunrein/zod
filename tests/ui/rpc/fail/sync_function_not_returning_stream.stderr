error[E0277]: the trait bound `std::string::String: futures_core::stream::Stream` is not satisfied
 --> tests/ui/rpc/fail/sync_function_not_returning_stream.rs:8:8
  |
8 |     fn test(&mut self) -> String {
  |        ^^^^ the trait `futures_core::stream::Stream` is not implemented for `std::string::String`
  |
  = help: the following other types implement trait `futures_core::stream::Stream`:
            &mut S
            Abortable<St>
            AssertUnwindSafe<S>
            Box<S>
            Buffer<S, Item>
            BufferUnordered<St>
            Buffered<St>
            Either<A, B>
          and $N others
note: required by a bound in `DependencyMap::add_stream_output`
 --> core/src/lib.rs
  |
  |     pub fn add_stream_output<F, S, T>(&mut self, _: F)
  |            ----------------- required by a bound in this associated function
...
  |         S: futures::Stream<Item = T>,
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `DependencyMap::add_stream_output`

error[E0277]: the trait bound `std::string::String: futures_core::stream::Stream` is not satisfied
 --> tests/ui/rpc/fail/sync_function_not_returning_stream.rs:6:1
  |
6 | #[rpc]
  | ^^^^^^ the trait `futures_core::stream::Stream` is not implemented for `std::string::String`
  |
  = help: the following other types implement trait `futures_core::stream::Stream`:
            &mut S
            Abortable<St>
            AssertUnwindSafe<S>
            Box<S>
            Buffer<S, Item>
            BufferUnordered<St>
            Buffered<St>
            Either<A, B>
          and $N others
note: required by a bound in `zod::zod_core::ast::Ref::new_stream_res`
 --> core/src/ast/schema/ref.rs
  |
  |     pub const fn new_stream_res<F, S, I>(_: &'static F) -> Self
  |                  -------------- required by a bound in this associated function
...
  |         S: futures::Stream<Item = I>,
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Ref::new_stream_res`
  = note: this error originates in the attribute macro `rpc` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `std::string::String: futures_core::stream::Stream` is not satisfied
   --> tests/ui/rpc/fail/sync_function_not_returning_stream.rs:8:8
    |
6   | #[rpc]
    | ------ required by a bound introduced by this call
7   | impl A {
8   |     fn test(&mut self) -> String {
    |        ^^^^ the trait `futures_core::stream::Stream` is not implemented for `std::string::String`
    |
    = help: the following other types implement trait `futures_core::stream::Stream`:
              &mut S
              Abortable<St>
              AssertUnwindSafe<S>
              Box<S>
              Buffer<S, Item>
              BufferUnordered<St>
              Buffered<St>
              Either<A, B>
            and $N others
    = note: required for `Pin<&mut std::string::String>` to implement `futures_core::stream::Stream`
note: required by a bound in `StreamExt::next`
   --> /home/nicolai/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/mod.rs:249:22
    |
249 | pub trait StreamExt: Stream {
    |                      ^^^^^^ required by this bound in `StreamExt::next`
...
273 |     fn next(&mut self) -> Next<'_, Self>
    |        ---- required by a bound in this associated function
